controllers/statusmanager/status_manager.go

```diff
@@ -5,64 +5,315 @@ package statusmanager
+		oldStatus := antreaInstall.Status.DeepCopy()
+		if reachedAvailableLevel {
+			co.Status.Versions = []configv1.OperandVersion{
+				{Name: "operator", Version: version.Version},
+			}
+		}
+		status.CombineConditions(&co.Status.Conditions, &conditions)
+		progressingCondition := v1helpers.FindStatusCondition(co.Status.Conditions, configv1.OperatorProgressing)
+		availableCondition := v1helpers.FindStatusCondition(co.Status.Conditions, configv1.OperatorAvailable)
+		if availableCondition == nil && progressingCondition != nil && progressingCondition.Status == configv1.ConditionTrue {
+			v1helpers.SetStatusCondition(&co.Status.Conditions,
+				configv1.ClusterOperatorStatusCondition{
+					Type:    configv1.OperatorAvailable,
+					Status:  configv1.ConditionFalse,
+					Reason:  "Startup",
+					Message: "The network is starting up",
+				},
+			)
+		}
+		v1helpers.SetStatusCondition(&co.Status.Conditions,
+			configv1.ClusterOperatorStatusCondition{
+				Type:   configv1.OperatorUpgradeable,
+				Status: configv1.ConditionTrue,
+			},
+		)
+		if reflect.DeepEqual(*oldStatus, co.Status) {
+			return nil
+		}
+		err = status.setAntreaInstallStatus(&co.Status.Conditions)
+		return err
+	})
+	if err != nil {
+		log.Error(err, "Failed to set AntreaInstall")
+	}
+}
+
+// Set updates the ClusterOperator.Status with the provided conditions for platform openshift.
+func (adaptor *StatusOc) set(status *StatusManager, reachedAvailableLevel bool, conditions ...configv1.ClusterOperatorStatusCondition) {
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		co := &configv1.ClusterOperator{ObjectMeta: metav1.ObjectMeta{Name: status.name}}
+		err := status.client.Get(context.TODO(), types.NamespacedName{Name: status.name}, co)
+		isNotFound := errors.IsNotFound(err)
+		if err != nil && !isNotFound {
+			return err
+		}
+		oldStatus := co.Status.DeepCopy()
+		if reachedAvailableLevel {
+			co.Status.Versions = []configv1.OperandVersion{
+				{Name: "operator", Version: version.Version},
+			}
+		}
+		status.CombineConditions(&co.Status.Conditions, &conditions)
+		progressingCondition := v1helpers.FindStatusCondition(co.Status.Conditions, configv1.OperatorProgressing)
+		availableCondition := v1helpers.FindStatusCondition(co.Status.Conditions, configv1.OperatorAvailable)
+		if availableCondition == nil && progressingCondition != nil && progressingCondition.Status == configv1.ConditionTrue {
+			v1helpers.SetStatusCondition(&co.Status.Conditions,
+				configv1.ClusterOperatorStatusCondition{
+					Type:    configv1.OperatorAvailable,
+					Status:  configv1.ConditionFalse,
+					Reason:  "Startup",
+					Message: "The network is starting up",
+				},
+			)
+		}
+		v1helpers.SetStatusCondition(&co.Status.Conditions,
+			configv1.ClusterOperatorStatusCondition{
+				Type:   configv1.OperatorUpgradeable,
+				Status: configv1.ConditionTrue,
+			},
+		)
+		if reflect.DeepEqual(*oldStatus, co.Status) {
+			return nil
+		}
```

Comment:
If possible, it might be worth avoiding duplication between code for the two adaptors

Reply:
Thanks Salvatore, this does seem not graceful and I'll have a try.